src/main/scala-2.13+/scala/concurrent/stm/skel/AtomicArray.scala

```diff
@@ -0,0 +1,281 @@
+      case _: Array[Char]    => new ofChar(size)
+      case _: Array[Int]     => new ofInt(size)
+      case _: Array[Float]   => new ofFloat(size)
+      case _: Array[Long]    => new ofLong(size)
+      case _: Array[Double]  => new ofDouble(size)
+      case _: Array[Unit]    => new ofUnit(size)
+      case _: Array[AnyRef]  => new ofRef[AnyRef](size)
+    }).asInstanceOf[AtomicArray[T]]
+  }
+
+  def apply(elems: Array[Boolean]) = new ofBoolean(new AtomicIntegerArray(elems map {if(_) 1 else 0}))
+  def apply(elems: Array[Byte])    = new ofByte(   new AtomicIntegerArray(elems map {_.toInt}))
+  def apply(elems: Array[Short])   = new ofShort(  new AtomicIntegerArray(elems map {_.toInt}))
+  def apply(elems: Array[Char])    = new ofChar(   new AtomicIntegerArray(elems map {_.toInt}))
+  def apply(elems: Array[Int])     = new ofInt(    new AtomicIntegerArray(elems))
+  def apply(elems: Array[Float])   = new ofFloat(  new AtomicIntegerArray(elems map {java.lang.Float.floatToRawIntBits(_)}))
+  def apply(elems: Array[Long])    = new ofLong(   new AtomicLongArray(elems))
+  def apply(elems: Array[Double])  = new ofDouble( new AtomicLongArray(elems map {java.lang.Double.doubleToRawLongBits(_)}))
+  def apply(elems: Array[Unit])    = new ofUnit(   elems.length)
+  def apply[T <: AnyRef](elems: Array[T]) =
+    new ofRef(new AtomicReferenceArray(elems.asInstanceOf[Array[AnyRef]]).asInstanceOf[AtomicReferenceArray[T]])
+
+  def apply[T](elems: IterableOnce[T])(implicit m: ClassTag[T]): AtomicArray[T] = {
+    val array: AnyRef = elems match {
+      case w: mutable.WrappedArray[_] => w.array // we're going to copy out regardless, no need to duplicate right now
+      case _ => elems.iterator.toArray
+    }
+    val result = array match {
+      case x: Array[Boolean]  => apply(x)
+      case x: Array[Byte]     => apply(x)
+      case x: Array[Short]    => apply(x)
+      case x: Array[Char]     => apply(x)
+      case x: Array[Int]      => apply(x)
+      case x: Array[Float]    => apply(x)
+      case x: Array[Long]     => apply(x)
+      case x: Array[Double]   => apply(x)
+      case x: Array[Unit]     => apply(x)
+      case x: Array[AnyRef]   => apply(x)
+    }
+    result.asInstanceOf[AtomicArray[T]]
+  }
+
+  final class ofBoolean(elems: AtomicIntegerArray) extends AtomicArray[Boolean] {
+    def this(size: Int) = this(new AtomicIntegerArray(size))
+
+    private def decode(v: Int) = v != 0
+    private def encode(elem: Boolean) = if (elem) 1 else 0
+
+    def length: Int = elems.length
+    def apply(index: Int): Boolean = decode(elems.get(index))
+    def update(index: Int, elem: Boolean): Unit = elems.set(index, encode(elem))
+    def swap(index: Int, elem: Boolean): Boolean = decode(elems.getAndSet(index, encode(elem)))
+    def compareAndSet(index: Int, expected: Boolean, elem: Boolean): Boolean =
+      elems.compareAndSet(index, encode(expected), encode(elem))
+    override def newBuilder = new AtomicArrayBuilder.ofBoolean
```

Comment:
I guess `newSpecificBuilder` should be overridden instead. (and that’s the case in all the other `AtomicArray` subclasses.